tests/test_frame.py

```diff
@@ -1,640 +1,183 @@
-        assert not events
-        assert len(navigatedFrames) == 1
-
-    @sync
-    async def test_frame_events_child(self):
-        attachedFrames = []
-        detachedFrames = []
-        navigatedFrames = []
-        self.page.on('frameattached', lambda f: attachedFrames.append(f))
-        self.page.on('framedetached', lambda f: detachedFrames.append(f))
-        self.page.on('framenavigated', lambda f: navigatedFrames.append(f))
-        await self.page.goto(self.url + 'assets/nested-frames.html')
-        assert len(attachedFrames) == 4
-        assert len(detachedFrames) == 0
-        assert len(navigatedFrames) == 5
-
-        attachedFrames.clear()
-        detachedFrames.clear()
-        navigatedFrames.clear()
-        await self.page.goto(self.url + 'empty')
-        assert len(attachedFrames) == 0
-        assert len(detachedFrames) == 4
-        assert len(navigatedFrames) == 1
-
-    @sync
-    async def test_frame_name(self):
-        await self.page.goto(self.url + 'empty')
-        await attachFrame(self.page, 'FrameId', self.url + 'empty')
-        await asyncio.sleep(0.1)
-        await self.page.evaluate(
-            '''(url) => {
-                const frame = document.createElement('iframe');
-                frame.name = 'FrameName';
-                frame.src = url;
-                document.body.appendChild(frame);
-                return new Promise(x => frame.onload = x);
-            }''',
-            self.url + 'empty',
-        )
-        await asyncio.sleep(0.1)
-
-        frame1 = self.page.frames[0]
-        frame2 = self.page.frames[1]
-        frame3 = self.page.frames[2]
-        assert frame1.name == ''
-        assert frame2.name == 'FrameId'
-        assert frame3.name == 'FrameName'
+from pyppeteer.errors import BrowserError
+from tests.utils import waitEvent, attachFrame, detachFrame, dumpFrames, navigateFrame
+
+
+@sync
+async def test_executionContext(isolated_page, server):
+    p = isolated_page
+    await p.goto(server.empty_page)
+    await attachFrame(p, server.empty_page, 'frame1')
+    assert len(p.frames) == 2
+
+    frame1, frame2 = p.frames
+    context1 = await frame1.executionContext
+    context2 = await frame2.executionContext
+    assert context1.frame is frame1
+    assert context2.frame is frame2
+
+    await gather(
+        context1.evaluate('window.a = 1'),
+        context2.evaluate('window.a = 2'),
+    )
+    a1, a2 = await gather(
+        context1.evaluate('window.a'),
+        context2.evaluate('window.a'),
+    )
+    assert a1 == 1
+    assert a2 == 2
+
+@sync
+async def test_evaluate(isolated_page, server):
+    p = isolated_page
+    # should throw for detached frames
+    frame1 = await attachFrame(p, server.empty_page, 'frame1')
+    await detachFrame(p, 'frame1')
+    with pytest.raises(BrowserError) as e:
+        await frame1.evaluate('7 * 8')
+    assert e.match('Execution Context is not available in detached frame')
```

Comment:
IMO the inline exception message matching is a little better:
```suggestion
    with pytest.raises(BrowserError, match='Execution Context is not available in detached frame') as e:
        await frame1.evaluate('7 * 8')
```